### PR TITLE
support plugin to skip parsing special extension instance.

### PR DESCRIPTION
**Pull request by fanchanghu**
Hi,
This PR is for the issue #1345 , however, it is a little more complex than I expected.

Please have a look and give some suggestions, thanks.  @rkrejci 

I will still work on this to add some tests if there is no major problems in this code.

### CODE REVIEW

File: src/parser.h [hunk 1]
```diff
@@ -162,6 +162,9 @@ const struct lys_node *lyp_get_yang_data_template(const struct lys_module *modul
 
 void lyp_ext_instance_rm(struct ly_ctx *ctx, struct lys_ext_instance ***ext, uint8_t *size, uint8_t index);
 
+int lyp_get_ext_list(struct ly_ctx *ctx, void *elem, LYEXT_PAR elem_type,
```
Comment (rkrejci) [inline src/parser.h:165]: Please add doxygen description with the short comment about the purpose of the function and its parameters.

File: src/parser_yang.c [hunk 2]
```diff
@@ -3334,26 +3334,50 @@ free_yang_common(struct lys_module *module, struct lys_node *node)
...
+        info->substmt_index = ext_list[i]->insubstmt_index;
         info->ext_index = i;
-        if (unres_schema_add_node(module, unres, ext, UNRES_EXT, (struct lys_node *)info) == -1) {
+
+        rc = unres_schema_add_node(module, unres, ext, UNRES_EXT, (struct lys_node *)info);
+        if (rc == -1) {
             return EXIT_FAILURE;
         }
+        if (!rc && !ext_list[i]) {
+            /* this extension is skipped, move all extensions after it */
```
Comment (rkrejci) [inline src/parser_yang.c:3362]: what about using `memmove(3)`?

File: src/parser_yin.c [hunk 3]
```diff
@@ -105,13 +106,16 @@ lyp_yin_fill_ext(void *parent, LYEXT_PAR parent_type, LYEXT_SUBSTMT substmt, uin
...
     info->substmt = substmt;
     info->substmt_index = substmt_index;
-    info->ext_index = ext_index;
+    info->ext_index = *ext_size;
 
-    if (unres_schema_add_node(module, unres, ext, UNRES_EXT, (struct lys_node *)info) == -1) {
-        return EXIT_FAILURE;
+    rc = unres_schema_add_node(module, unres, ext, UNRES_EXT, (struct lys_node *)info);
+    if (!rc && !(*ext)[*ext_size]) {
+        // extension instance is skipped
```
Comment (rkrejci) [inline src/parser_yin.c:113]: please use block comments `/* */`

File: src/resolve.c [hunk 4]
```diff
@@ -7342,6 +7361,69 @@ print_unres_schema_item_fail(void *item, enum UNRES_ITEM type, void *str_node)
     }
 }
 
+static int
+save_ignored_ext_parent(int resolve_rc, struct ly_set *ext_parents, struct ly_set *ext_par_types,
```
Comment (rkrejci) [inline src/resolve.c:7365]: please add doxygen description comment about the function and its parameters

File: src/resolve.c [hunk 5]
```diff
@@ -7342,6 +7361,69 @@ print_unres_schema_item_fail(void *item, enum UNRES_ITEM type, void *str_node)
...
+
+    finish:
+    if(!resolve_rc || resolve_rc == -1) {
+        free_ext_data(ctx, ext_data);
+    }
+    return rt;
+}
+
+static int
+compack_ext_list(struct ly_ctx *ctx, void *elem, LYEXT_PAR elem_type)
```
Comment (rkrejci) [inline src/resolve.c:7390]: please add doxygen description comment about the function and its parameters

File: src/resolve.c [hunk 6]
```diff
@@ -7419,6 +7520,17 @@ resolve_unres_schema_types(struct unres_schema *unres, enum UNRES_ITEM types, st
...
 
+    if ((types & UNRES_EXT)) {
+        assert(ext_parents->number == ext_par_types->number);
+        for (i = 0; i < ext_parents->number; ++i) {
+            ret = compack_ext_list(ctx, ext_parents->set.g[i], (LYEXT_PAR)ext_parents->set.g[i]);
+            LY_CHECK_GOTO(ret == -1, finish);
+        }
+    }
+
+    finish:
```
Comment (rkrejci) [inline src/resolve.c:7531]: formatting - move the label to the column 0

**Review by rkrejci — CHANGES_REQUESTED** [review]

**Comment by rkrejci** [comment]
Could you please also, here or in the referenced issue, a little elaborate on why the change is more complex than we've expected? I agree that the test will be needed, but except for the minor issues I've pointed out, it seems fine to me.

**Comment by fanchanghu** [comment]
Hi,
The review issues have been fixed, please have a look again, thanks.

I only thought of processing the skipped extension instance when parsing it at first, however, when I worked on this, I found that it also need be processed when resolved. And there could be more than 1 skipped extension instances when resolving, so I have to saving their parents in a set and compact the extension instance arrays later. That's why I said it was a little complex than I expected.

**Comment by rkrejci** [comment]
ok, the code looks fine to me now, are you going to add some test cases?

**Review by michalvasko — APPROVED** [review]

**Comment by fanchanghu** [comment]
Yes, I'm working on the test cases now, the test cases will be pushed later, with a memory leak issue (found by the test cases) patch.

**Comment by fanchanghu** [comment]
Hi,
The test cases and memory leak issue patch have been committed, please check it, thanks.

**Review by rkrejci — APPROVED** [review]

**Pull request merged**